localization/kitti_evaluator/src/kitti_evaluator_core.cpp

```diff
@@ -0,0 +1,210 @@
+    ground_truth_pose_topic, rclcpp::QoS{100},
+    std::bind(&KittiEvaluator::callbackGroundTruthOdometry, this, std::placeholders::_1));
+  pose_error_pub_ = create_publisher<geometry_msgs::msg::PoseStamped>(error_pose_topic, 1);
+  init = false;
+  last_vehicle_trans.setIdentity();
+  last_groud_truth_trans.setIdentity();
+}
+
+KittiEvaluator::~KittiEvaluator() {}
+
+void KittiEvaluator::initialize(double lat, double lon, double alt)
+{
+  init = true;
+  scale = cos(lat * geodetic::pi / 180.0);
+  std::vector<double> t = geodetic::se3_translation(lat, lon, alt, scale);
+  x0 = t[0];
+  y0 = t[1];
+  z0 = t[2];
+}
+
+// void KittiEvaluator::callbackNavSatFix(sensor_msgs::msg::NavSatFix::ConstSharedPtr
+// nav_sat_fix_msg_ptr)
+// {
+//     RCLCPP_INFO(get_logger(), "Get GPS");
+//     if(!init)
+//     {
+//         initialize(nav_sat_fix_msg_ptr->latitude, nav_sat_fix_msg_ptr->longitude,
+//         nav_sat_fix_msg_ptr->altitude);
+//     }
+//     auto curr_time = nav_sat_fix_msg_ptr->header.stamp;
+//     while(!odom_queue.empty() && rclcpp::Time(odom_queue.front()->header.stamp) <
+//     rclcpp::Time(curr_time) - rclcpp::Duration::from_seconds(delta))
+//     {
+//         odom_queue.pop_front();
+//     }
+//     if(!odom_queue.empty())
+//     {
+//         // calculateError(odom_queue.front(), nav_sat_fix_msg_ptr);
+//     }
+//     else
+//     {
+//         RCLCPP_INFO(get_logger(), "Cannot Find Valid Pose!");
+//     }
+// }
+
+void KittiEvaluator::callbackGroundTruthOdometry(
+  geometry_msgs::msg::PoseStamped::ConstSharedPtr pose_ground_truth_msg_ptr)
+{
+  if (!vehicle_pose_queue.empty()) {
+    RCLCPP_INFO(get_logger(), "We have ekf pose! Size : %ld", vehicle_pose_queue.size());
+  }
+
+  Eigen::Affine3d ground_truth_trans;
+  msgToMatrix(pose_ground_truth_msg_ptr, ground_truth_trans);
+
+  rclcpp::Time curr_time(pose_ground_truth_msg_ptr->header.stamp);
+  geometry_msgs::msg::PoseStamped::ConstSharedPtr prev_pose;
+  while (!vehicle_pose_queue.empty() && rclcpp::Time(vehicle_pose_queue.front()->header.stamp) <
+                                          curr_time - rclcpp::Duration::from_seconds(delta)) {
+    prev_pose = vehicle_pose_queue.front();
+    vehicle_pose_queue.pop_front();
+  }
+  if (!vehicle_pose_queue.empty()) {
+    auto post_pose = vehicle_pose_queue.front();
+    vehicle_pose_queue.pop_front();
+
+    Eigen::Affine3d prev_trans;
+    rclcpp::Time prev_time = get_clock()->now();
+    prev_trans.setIdentity();
+    if (prev_pose) {
+      prev_time = rclcpp::Time(prev_pose->header.stamp);
+      msgToMatrix(prev_pose, prev_trans);
+    }
+    Eigen::Affine3d post_trans;
+    rclcpp::Time post_time(post_pose->header.stamp);
+    msgToMatrix(post_pose, post_trans);
+
+    rclcpp::Time fake_time = curr_time - rclcpp::Duration::from_seconds(delta);
+    double t = interpolation::getTimeCoeffients(fake_time, prev_time, post_time);
+    if (t > 1 || t < 0) {
+      RCLCPP_INFO(get_logger(), "????????????");
```

Comment:
This should have proper output message. We might want to consider skipping the rest of the process if there is something wrong with the interpolation time.